content/courses/Ada_For_The_C_Embedded_Developer/chapters/07_Reusability.rst

```diff
@@ -497,10 +497,270 @@ Object orientation
+we could have a two-layered system:
+
+- A layer of the system (an event manager) triggers events depending on
+  information from sensors.
+
+    - For each event, callback functions can be registered.
+
+    - The event manager calls registered callback functions when an event is
+      triggered.
+
+- Another layer of the system registers callback functions for specific events
+  and decides what to do when those events are triggered.
+
+This approach promotes information hiding and component decoupling because:
+
+- the layer of the system responsible for managing events doesn't need to know
+  what the callback function actually does, while
+
+- the layer of the system that implements callback functions remains
+  agnostic to implementation details of the event manager |mdash| for example,
+  how events are implemented in the event manager.
+
+Let's see an example in C where we have a :c:`process_values()` function that
+calls a callback function (:c:`routine`) to process a list of values:
+
+[C]
+
+.. code:: c manual_chop run_button project=Courses.Ada_For_C_Embedded_Dev.Reusability.Callback_Routine_C
+
+    !process_values.h
+    typedef int (*callback_routine) (int);
+
+    void process_values (int              *values,
+                         int               len,
+                         callback_routine  routine);
+
+    !process_values.c
+    #include "process_values.h"
+
+    #include <assert.h>
+
+    void process_values (int              *values,
+                         int               len,
+                         callback_routine  routine)
+    {
+        int i;
+
+        assert (routine != null);
+
+        for (i = 0; i < len; i++)
+        {
+            values[i] = routine (values[i]);
+        }
+    }
+
+    !main.c
+    #include <stdio.h>
+    #include <stdlib.h>
+
+    #include "process_values.h"
+
+    int proc_10 (int val)
+    {
+        return val + 10;
+    }
+
+    # define LEN_VALUES     5
+
+    int main()
+    {
+
+        int values[LEN_VALUES] = { 1, 2, 3, 4, 5 };
+        int i;
+
+        process_values (values, LEN_VALUES, &proc_10);
+
+        for (i = 0; i < LEN_VALUES; i++)
+        {
+            printf("Value [%d] = %d\n", i, values[i]);
+        }
+    }
+
+As mentioned previously, :c:`process_values()` doesn't have any knowledge about
+what :c:`routine()` does with the integer value it receives as a parameter.
+Also, we could replace :c:`proc_10()` by another function without having to
+change the implementation of :c:`process_values()`.
+
+Note that :c:`process_values()` calls an :c:`assert()` for the function
+pointer to compare it against :c:`null`. Here, instead of checking the validity
+of the function pointer, we're expecting the caller of :c:`process_values()`
+to provide a valid pointer.
+
+This is the corresponding implementation in Ada:
+
+[Ada]
+
+.. code:: ada run_button project=Courses.Ada_For_C_Embedded_Dev.Reusability.Callback_Routine_Ada
+
+    package Process_Values_Pkg is
```

Comment:
The "*_Pkg" suffix has the same problems as "*_T" -- how about "Values_Processing" for the package name?

Reply:
Thanks, I tried to, but couldn't come up with a better name. `Values_Processing` is perfect.